File: iarmud/world.rb
# Define the different exits available
class Symbol
	def complete
		case self.to_s.downcase[0]
			when ?n; :north
			when ?s; :south
			when ?e; :east
			when ?w; :west
			when ?u; :up
			when ?d; :down
			else; self
		end
	end

	def opposite
		case self
			when :north; :south
			when :south; :north
			when :east; :west
			when :west; :east
			when :up; :down
			when :down; :up
			when :northeast; :southwest
			when :southeast; :northwest
			when :southwest; :northeast
			when :northwest; :southeast
		end
	end

	def position(x,y)
		case self
			when :north; [x, y - 1]
			when :south; [x, y + 1]
			when :east; [x + 1, y]
			when :west; [x - 1, y]
			when :northeast; [x + 1, y - 1]
			when :southeast; [x + 1, y + 1]
			when :southwest; [x - 1, y + 1]
			when :northwest; [x - 1, y - 1]
		end
	end
end

# Standard directions
Directions = [:north, :south, :east, :west, :northeast, :southeast, :southwest, :northwest]
All_Directions = [:north, :south, :east, :west, :northeast, :southeast, :southwest, :northwest, :up, :down]

# A standard room
class Room
	def initialize (id, title, description = "", exits = {}, tile = " ")
		@id, @title, @description = id, title, description 
		@exits = exits
		@tile = tile
		@art = {}
		@contents = []
	end

	def join (dir, room)
		@exits[dir] = room
		room.exits[dir.opposite] = self
	end

	def border (dir)
		return (@art[dir] || ".") unless @exits[dir]
		return @exits[dir].tile
	end

	def players
		@contents.find_all { |i| i.is_a? Player }
	end

	def contents_as_s(exclude = nil)
		ret = ""
		@contents.each { |i| 
			next if (i.name == exclude) || i.class == exclude
			if i.is_a? Player
				ret += "#{i.name} is here.\n"
			else
				ret += "There is #{i.name} here.\n"
			end
		}
		ret
	end

	def move_entity (what, dir)
		if @exits[dir]
			@contents.delete what
			@exits[dir].contents.push what
			return true
		end
		false
	end

	def map
		v = WorldVisualizer.new(self)
		v.visualize
	end

	attr_accessor :id, :title, :description, :contents, :exits, :tile, :art
end

class WorldVisualizer
	def initialize (start)
		@root = start
		@min_x, @min_y, @max_x, @max_y = -1, -1, 1, 1
		@visual = {}
		@done = {}
	end

	def visualize
		map_room @root, [0, 0]
		size_x = -@min_x + @max_x
		size_y = -@min_y + @max_y
		map = []
		@min_y.upto(@max_y) do |y|
			line = []
			@min_x.upto(@max_x) do |x|
				line.push @visual[[x, y]] || " "
			end
			map.push line.join
		end
		return map.join("\n")
	end

	def map_room (room, pos, prev = nil)
		return if @done[pos]
		@done[pos] = true
		prev = room if prev == nil
		x, y = pos[0], pos[1]
		@min_x = x - 1 if x - 1 < @min_x
		@min_y = y - 1 if y - 1 < @min_y
		@max_x = x + 1 if x + 1 > @max_x
		@max_y = y + 1 if y + 1 > @max_y

		if room.contents.length > 0
			@visual[[x, y]] = room.contents.length
		else
			@visual[[x, y]] = room.tile
		end
		Directions.each do |dir|
			pos = dir.position(x, y)
			@visual[pos] = room.border(dir) unless @visual[pos]
		end

		Directions.each do |dir|
			if room.exits[dir]
				map_room room.exits[dir], dir.position(x, y), prev unless prev == room.exits[dir]
			end
		end
	end
end

class World
	def initialize
		@root = Room.new(0, "Spawning room", "Spawning vat")
		@next_id = 1
	end

	def create_room (title, description = nil, exits = {}, tile = " ")
		id = @next_id
		@next_id++
		Room.new(id, title, description, exits, tile)
	end

	def map
		v = WorldVisualizer.new(@root)
		v.visualize
	end

	attr_accessor :root, :next_id
end

#this is a base class for any Item which a Location can contain:
#	any object in our Game world, from players to AIs to swords and gold.
# you differentiate items in the game world based on their class.
# it doesn't have much in the way of attrubutes: a reference to its location, 
#	a name and description
class Mappable
	attr_accessor :name, :description, :location
	def initialize(name, description, location = nil)
		@name, @description, @location = name, description, location
	end
end

#this is the base class for all living creatures: Human players, AIs
# it has attributes like: health, race, strength, dexterity, speed
class Creature < Mappable
	attr_reader :title, :race, :health, :armour
	def initialize(name, title = nil, race = nil)
		@title = title ? title : ""
		@race = race ? race : "Human"
		@health = 100
		@armour = 0
		super(name, "A #{race}")
	end
end

#this is something which goes in the world which is not a player.
#	it has things like: weight (can a player pick it up? should it just be 'carryable'?)
class Item < Mappable
	
end

# Example subclass:
# swords, knives...
class StabbyItem < Item # < AttackItem ?
	
end

#this is a base class for events which occur within the world.
#	they have a location. Multiple events can be chained into a 
#	sequence.
# we might want a duration / start time (in ticks) for events or
#	a descendant thereof - earthquakes take more than one tick. 
#		maybe so does an attack, leaving you vulnerable...
class Event
	attr_accessor :text, :before, :after
	def initialize(text)
		@text = text
		@before = @after = nil
	end
	
	def insert_before(event)
		event = Event.new(event) if !event.is_a? Event
		@before.after = event if @before
		@before = event
		event.after = self
	end
	
	def insert_after(event)
		event = Event.new(event) if !event.is_a? Event
		@after.before = event if @after
		@after = event
		event.before = self
	end
	
	def to_s
		@text
	end
	
end

class FailureEvent < Event
	def to_s
		"EPIC FAIL: #{@text}"
	end
end

def testworld
	w = World.new
	w.root.join(:east, Room.new(w.next_id + 1, "East", "East"))
	w.root.join(:south, Room.new(w.next_id + 2, "South", "South"))
	w.root.join(:north, Room.new(w.next_id + 3, "North", "North"))
	w.root.join(:west, Room.new(w.next_id + 4, "West", "West"))
	w.root.join(:northeast, Room.new(w.next_id + 5, "Northeast", "NE"))
end

File: src/world.erl
-module(world).

-export([room/5, map/2, join/3, test/0]).

-define(directions, [north, south, east, west, northeast, southeast, southwest,
	northwest]).
-define(all_directions, ?directions ++ [up, down]).

opposite(north) -> south;
opposite(south) -> north;
opposite(east) -> west;
opposite(west) -> east;
opposite(northeast) -> southwest;
opposite(southeast) -> northwest;
opposite(southwest) -> northeast;
opposite(northwest) -> southeast;
opposite(What) -> error({badarg, What}).

-type pos() :: {integer(), integer()}.

position(Pos, {X, Y}) ->
	case Pos of
		north -> {X, Y - 1};
		south -> {X, Y + 1};
		east  -> {X + 1, Y};
		west  -> {X - 1, Y};
		northeast -> {X + 1, Y - 1};
		southeast -> {X + 1, Y + 1};
		southwest -> {X - 1, Y + 1};
		northwest -> {X - 1, Y - 1};
		_ -> error({badarg, Pos})
	end.

-type direction() :: north | south | east | west | northeast | southeast | southwest | northwest.
-type room_id() :: reference().

-record(exit, {
	direction :: direction(),
	room_id :: room_id(),
	tile :: binary()
}).
-type exit() :: #exit{}.

-record(room, {
	id    :: room_id(),
	title :: binary(),
	description = <<"">> :: binary(),
	exits = [] :: [exit()],
	tile = <<" ">> :: binary(),
	art = [] :: [{direction(), binary()}],
	contents = [] :: [term()]
}).
-type room() :: #room{}.

room(Id, Title, Description, Exits, Tile) ->
	#room{
		id = Id,
		title = Title,
		description = Description,
		exits = Exits,
		tile = Tile
	}.

join(Direction, #room{ id = Id, tile = Tile }, #room{ exits = Exits0 } = Room) ->
	Exits1 = [#exit{
		direction = Direction,
		room_id = Id,
		tile = Tile
	} | lists:keydelete(Direction, #exit.direction, Exits0)],
	Room#room{
		exits = Exits1
	}.

border(Direction, #room{ art = Art, exits = [] }) ->
	case lists:keyfind(Direction, 1, Art) of
		false -> <<".">>;
		{_, Tile} -> Tile
	end;
border(Direction, #room{ art = Art, exits = Exits }) ->
	case lists:keyfind(Direction, #exit.direction, Exits) of
		#exit{ tile = Tile } -> Tile;
		false -> border(Direction, #room{ art = Art })
	end.

map(#room{} = Room, Rooms) ->
	V = vis(Room, Rooms),
	visualize(V).

-record(vis, {
	rooms = [] :: {room_id(), room()},
	root :: room(),
	min_x = -1,
	min_y = -1,
	max_x =  1,
	max_y =  1,
	visual = [] :: [{pos(), binary()}],
	done = [] :: [{pos(), done}]
}).
-type vis() :: #vis{}.

vis(Room, Rooms) ->
	#vis{
		rooms = Rooms,
		root = Room
	}.

visualize(#vis{} = Vis0) ->
	Vis1 = map_room(Vis0#vis.root, {0, 0}, Vis0),
	XSeq = lists:seq(Vis1#vis.min_x, Vis1#vis.max_x),
	YSeq = lists:seq(Vis1#vis.min_y, Vis1#vis.max_y),
	[ begin
		[ case lists:keyfind({X, Y}, 1, Vis1#vis.visual) of
			false -> <<".">>;
			{_, Tile} -> Tile
		  end || X <- XSeq ] ++ <<"\n">>
	  end || Y <- YSeq ].

map_room(#room{} = Room, Pos, #vis{} = Vis) ->
	map_room(Room, Pos, Vis, nil).

map_room(#room{} = Room, Pos, #vis{} = Vis, Prev) ->
	IsDone = lists:keyfind(Pos, 1, Vis#vis.done) == {Pos, done},
	maybe_map_room(IsDone, Room, Pos, Vis, Prev).

-define(if_true(Test, New, Old), if Test -> New; true -> Old end).
maybe_map_room(true,  _,    _,            Vis, _) ->
	Vis;
maybe_map_room(false, Room, {X, Y} = Pos, #vis{ visual = Visual0} = Vis0, Prev0) ->
	io:format("Prev0: ~p~n", [Prev0]),
	Prev1 = if
		Prev0 == nil -> Room#room.id;
		true -> Prev0
	end,
	io:format("Prev: ~p~n", [Prev1]),
	Tile = if
		length(Room#room.contents) > 0 -> list_to_binary(integer_to_list(length(Room#room.contents)));
		true -> Room#room.tile
	end,
	Visual1 = [{Pos, Tile} | Visual0],
	Visual2 = lists:foldl(fun(Direction, VisualAcc) ->
		Position = position(Direction, Pos),
		case lists:keyfind(Position, 1, VisualAcc) of
			false ->
				case lists:keyfind(Direction, #exit.direction, Room#room.exits) of
					false -> VisualAcc;
					Exit ->
						Rm = lists:keyfind(Exit#exit.room_id, #room.id, Vis0#vis.rooms),
						[{Position, border(Direction, Rm)} | VisualAcc]
				end;
			_ ->
				VisualAcc
		end
	end, Visual1, ?directions),
	Vis1 = Vis0#vis{
		done = [{Pos, done} | Vis0#vis.done],
		min_x = ?if_true(X - 1 < Vis0#vis.min_x, X - 1, Vis0#vis.min_x),
		min_y = ?if_true(Y - 1 < Vis0#vis.min_y, Y - 1, Vis0#vis.min_y),
		max_x = ?if_true(X + 1 > Vis0#vis.max_x, X + 1, Vis0#vis.max_x),
		max_y = ?if_true(Y + 1 > Vis0#vis.max_y, Y + 1, Vis0#vis.max_y),
		visual = Visual2
	},
	lists:foldl(fun(Direction, VisAcc) ->
		case lists:keyfind(Direction, #exit.direction, Room#room.exits) of
			#exit{ room_id = Id } when Id /= Prev1 ->
				Rm = lists:keyfind(Id, #room.id, Vis1#vis.rooms),
				Position = position(Direction, Pos),
				map_room(Rm, Position, VisAcc, Prev1);
			_ ->
				VisAcc
		end
	end, Vis1, ?directions).

test() ->
	R1 = room(make_ref(), <<"Start">>, <<>>, [], <<" ">>),
	R2 = room(make_ref(), <<"East">>, <<>>, [], <<" ">>),
	R3 = room(make_ref(), <<"South">>, <<>>, [], <<" ">>),
	R1a = join(east, R2, R1),
	R2a = join(south, R3, R2),
	R2b = join(west, R1, R2a),
	R3a = join(north, R2, R3),
	Rs = [R1a, R2b, R3a],
	io:format("~s~n", [map(R1a, Rs)]).
